src/VecSim/spaces/IP/IP_SSE3_BF16.h

```diff
@@ -58,52 +58,47 @@ float BF16_InnerProductSIMD32_SSE3(const void *pVect1v, const void *pVect2v, siz
     __m128 sum_prod = _mm_setzero_ps();
 
     // handle first residual % 8 elements (smaller than step chunk size)
-    unsigned char skip = 0;
-    if (residual % 8) {
+    // | bf16_0 | bf16_1 | bf16_2 | bf16_3 |
+    if constexpr (residual % 8 >= 4) {
+        __m128i v1 = _mm_lddqu_si128((__m128i *)pVect1);
+        __m128i v2 = _mm_lddqu_si128((__m128i *)pVect2);
+        InnerProductLowHalfStep(v1, v2, _mm_setzero_si128(), sum_prod);
+        pVect1 += 4;
+        pVect2 += 4;
+    }
```

Comment:
IMO having the residual from small to large is clearer, consider moving to after the %4 block and before the extra steps.